Kernel/Term.hpp

```diff
@@ -271,14 +271,26 @@ class Term
 {
 public:
   //special functor values
-  static const unsigned SF_ITE = 0xFFFFFFFF;
-  static const unsigned SF_LET = 0xFFFFFFFE;
-  static const unsigned SF_FORMULA = 0xFFFFFFFD;
-  static const unsigned SF_TUPLE = 0xFFFFFFFC;
-  static const unsigned SF_LET_TUPLE = 0xFFFFFFFB;
-  static const unsigned SF_LAMBDA = 0xFFFFFFFA;
-  static const unsigned SF_MATCH = 0xFFFFFFF9;
-  static const unsigned SPECIAL_FUNCTOR_LOWER_BOUND = 0xFFFFFFF9;
+  enum class SpecialFunctor {
+    ITE,
+    LET,
+    FORMULA,
+    TUPLE,
+    LET_TUPLE,
+    LAMBDA,
+    MATCH, // <- keep this one the last, or modify SPECIAL_FUNCTOR_LAST accordingly
+  };
+  static constexpr SpecialFunctor SPECIAL_FUNCTOR_LAST = SpecialFunctor::MATCH;
+
+  static constexpr unsigned SPECIAL_FUNCTOR_LOWER_BOUND  = 0xFFFFFFFF - unsigned(SPECIAL_FUNCTOR_LAST);
```

Comment:
I guess this is packed into a bitfield at some point? If so can the magic constant be pulled out as a definition like `FUNCTOR_BITS` so that `SPECIAL_FUNCTOR_LOWER_BOUND = (1 << FUNCTOR_BITS) - 1` or something?

Reply:
I'm not quiiite sure what you mean here..
What we have right not (same as before) is that we just have `0..SPECIAL_FUNCTOR_LOWER_BOUND - 1` are normal functors, and the rest is reserved for the special ones. There's no bit-fields involved here.

Reply:
Sorry, I should have been clearer. Why `0xFFFFFFFF` - where does this number come from? It makes me concerned that this happens to work because you compiled it on something where the integers are the right size.

Reply:
Very true. I think this magic constant was put there by someone to represent the max of usize originally, and I copied it. I was gonna replace it by `std::numeric_limits<unsigned>::max()`, which I did in some places but forgot this one apparently.

Reply:
Fixed.